Create HIP-859 for adding `gas_consumed` field (#859)
hashgraph/hedera-improvement-proposal: HIP/hip-859.md

---
hip: 859
title: Add support for gasConsumed field in REST APIs related to contract results
author: Ivan Kavaldzhiev <ivan.kavaldzhiev@limechain.tech>, Stoyan Panayotov <stoyan.panayotov@limechain.tech>
working-group: Steven Sheehy <steven.sheehy@swirldslabs.com>, Nana Essilfie-Conduah <nana@swirldslabs.com>, Danno Ferrin <danno.ferrin@swirldslabs.com>, David Bakin <david.bakin@swirldslabs.com>
type: Standards Track
category: Mirror
needs-council-approval: Yes
status: Draft
created: 17.01.2024
discussions-to: https://github.com/hashgraph/hedera-improvement-proposal/discussions/860
---

## Abstract

> Currently, consensus nodes export a `gasUsed` field for each EVM impacting transaction, as part of the record streaming. As a result, the `contract_result` table in the mirror node database contains a `gas_used` field for each transaction. This field value is calculated using the gas refund policy applied in the consensus node.

Reviewer: ```suggestion
Currently, consensus nodes export a `gasUsed` field for each EVM impacting transaction and the mirror node persists and exposes this as `gas_used` on the contract result REST APIs. This field value is calculated using the gas refund policy applied in the consensus node.
```

Author: Fixed

This means that the value in this field represents how much gas the user was actually charged, not the amount of gas that the EVM consumed during the execution. This value can greatly differ from the actual amount that was used by the EVM. 

Let's say we use a 20% refund policy for the `gasLimit` that is sent by the user.

That means if the EVM consumes for example 55,500 gas, but the user is passing 400,000 gas limit, they will be charged the `max(55,500, 320,000)`, so it would be 320,000. That is the value calculated after we return the 20% of the gasLimit back to the user, which is 80,000.
The logic compares the bigger of the two values - the gas consumed by the EVM and the 80% of the `gasLimit` that the user has passed.

In this way we lose track of the actual gas that was consumed by the EVM.

> This HIP will add a new field `gas_consumed` to the `contract_result` table in the mirror node, which will contain the actually consumed gas by the EVM.

Author: Fixed

## Motivation

> Adding a `gas_consumed` field in the `contract_result` table would enrich user experience by providing more information about the transaction. 

Author: Fixed

---
hip: 859
title: Add support for gasConsumed field in REST APIs related to contract results
author: Ivan Kavaldzhiev <ivan.kavaldzhiev@limechain.tech>, Stoyan Panayotov <stoyan.panayotov@limechain.tech>
working-group: Steven Sheehy <steven.sheehy@swirldslabs.com>, Nana Essilfie-Conduah <nana@swirldslabs.com>, Danno Ferrin <danno.ferrin@swirldslabs.com>, David Bakin <david.bakin@swirldslabs.com>
type: Standards Track
category: Mirror
needs-council-approval: Yes
status: Draft
created: 2024-01-17
discussions-to: https://github.com/hashgraph/hedera-improvement-proposal/discussions/860
updated: 2024-01-22
---

## Abstract

> Currently, consensus nodes export a `gasUsed` field for each EVM impacting transaction and the mirror node persists and exposes this as `gas_used` on the contract result REST APIs. This field value is calculated using the gas refund policy applied in the consensus node.

Reviewer 2: nit: Might be nice to list the transactions. Not all readers may be aware of the 3 types at the time f writing.
```suggestion
Currently, consensus nodes export a `gasUsed` field for each EVM impacting transaction ([ContractCall](https://github.com/hashgraph/hedera-protobufs/blob/v0.46.0/services/contract_call.proto#L43), [ContractCreate](https://github.com/hashgraph/hedera-protobufs/blob/v0.46.0/services/contract_create.proto#L88) and [EthereumTransaction](https://github.com/hashgraph/hedera-protobufs/blob/v0.46.0/services/ethereum_transaction.proto#L31)) and the mirror node persists and exposes this as `gas_used` on the contract result REST APIs. This field value is calculated using the gas refund policy applied in the consensus node.
```

Author: Added

This means that the value in this field represents how much gas the user was actually charged, not the amount of gas that the EVM consumed during the execution. This value can greatly differ from the actual amount that was used by the EVM. 

Let's say we use a 20% refund policy for the `gasLimit` that is sent by the user.

> That means if the EVM consumes for example 55,500 gas, but the user is passing 400,000 gas limit, they will be charged the `max(55,500, 320,000)`, so it would be 320,000. That is the value calculated after we return the 20% of the gasLimit back to the user, which is 80,000.

Reviewer 3: I would maybe consider writing this as `max(55500, 0.8 * 400000)`.   ???

Author: Good idea, I edited the example

The logic compares the bigger of the two values - the gas consumed by the EVM and the 80% of the `gasLimit` that the user has passed.

In this way we lose track of the actual gas that was consumed by the EVM.

This HIP will add a new field `gas_consumed` to the mirror node contract result REST APIs, which will contain the actually consumed gas by the EVM.

## Motivation

Adding a `gas_consumed` field to the contract result REST APIs would enrich user experience by providing more information about the transaction. 

> This would allow users to better understand the gas refund policy and the actual gas consumption of their transactions. This would also allow users to understand how heavy their executed transaction was for the EVM.

Reviewer 3: Personally I would exchange the order of "actual gas consumption of their transactions" and "better understand the gas refund policy" because I believe that _most_ use of this will be by contract developers looking to optimize their contracts (point #1 in "user stories" below).

Author: Fixed

## Rationale

Consensus nodes export ContractActions sidecar records for each transaction after the introduction of [HIP-513](./hip-513.md). Each sidecar record contains a `gasUsed` field for the given action that the record represents (different type of call or contract creation).
This value directly represents the gas consumed by the EVM, without taking into account the gas refund policy. This makes it possible to calculate the overall gas consumption for the transaction, during the ContractActions sidecar record importing.  

> If we sum up all of the `gasUsed` values of each ContractActions sidecar for a given transaction on top of the initial intrinsic gas, which is the initial gas, charged prior to the transaction execution, then we would get accurate value for the `gasConsumed` field.

Reviewer 3: typo: "initial gas, charged prior" -> "initial gas charged prior"

Author: Fixed

## User stories

1. As a user, I want to see how much gas was consumed by the EVM during the execution of my transaction, when I retrieve a transaction info from the mirror-node REST APIs, so that I know if I can optimise my transaction cost
2. As a user, I want to see and compare how much gas I was charged for and how much gas was consumed by the EVM, when I inspect my transaction in hashscan

## Specification

In order to calculate the `gasUsed` field, the mirror node should have `CONTRACT_ACTION` and `CONTRACT_BYTECODE` sidecars enabled. Then, we would use an enhancement logic in the `mirror-node-importer`,
which will iterate over all `ContractAction` sidecar records for the transaction we want to import and calculate the `gasConsumed` field for.

The new logic would have a mechanism to first calculate the intrinsic gas cost for the transaction being imported. We can inspect if the transaction has a `ContractBytecode` sidecar

> related to it. If yes, we have a contract creation transaction and we would calculate the intrinsic gas cost based on the length of the init bytecode. If we don't have such a sidecar,

Reviewer 3: Consider changing the "if yes"/"if no" to bullet points?

Author: Fixed

then the transaction represents a contract call and we would use the default intrinsic gas cost for contract calls, which is 21,000.

The same logic could be applied also to `EthereumTransactions`, since they wrap the same two categories - contract create and contract call. If we have a `ContractBytecode` sidecar for an `EthereumTransaction`, then it wraps a contract create,
if not, it wraps a contract call.

On top of the intrinsic gas cost, we would iterate over all `ContractAction`
sidecars and sum up the `gasUsed` field for each of them. The result would be the `gasConsumed` field for the transaction.

For historical contract results we have the following options:

1. For the contract transactions that were executed after we have started exporting sidecars from consensus nodes, we can make a DB migration in mirror node and populate the `gasConsumed` field for them.
2. For the contract transactions that were executed prior to the introduction of sidecars or were executed after this period, but have missing sidecar records, we can put a `null` value for the `gasConsumed` field. In this case it would mean the value is not applicable and missing.

```

> Example:

Reviewer 3: Perhaps the formatting could be changed so this entire block isn't "code" style, just the equations at L71 and L75

Author: Fixed

1. The transaction is a ContractCall

gas_consumed = intrinsic_gas + sum(contract_action_gas_used)

2. The transaction is a ContractCreate

gas_consumed = intrinsic_gas + cost(contract_init_bytecode) + sum(contract_action_gas_used)
```

The `gas_used` field in `contract_result` table would still have the same semantics and represent the gas that was charged to the user.

The following REST APIs will be extended with the new field:

```
1. GET /api/v1/contracts/results
2. GET /api/v1/contracts/results/{transactionIdOrHash}
3. GET /api/v1/contracts/{contractIdOrAddress}/results 
4. GET /api/v1/contracts/{contractIdOrAddress}/results/{timestamp}
```

After the change, the API response will look like this:

```json
{
  "error_message": "Out of gas",
  "failed_initcode": "0x856739",
  "from": "0000000000000000000000000000000000001f41",
  "function_parameters": "0xbb9f02dc6f0e3289f57a1f33b71c73aa8548ab8b",
  "gas_consumed": 55500,

>   "gas_limit": 100000,

Reviewer 4: ```suggestion
  "gas_limit": 400000,
```

This means that the value in this field represents how much gas the user was actually charged, not the amount of gas that the EVM consumed during the execution. This value can greatly differ from the actual amount that was used by the EVM. 

Let's say we use a 20% refund policy for the `gasLimit` that is sent by the user.

That means if the EVM consumes for example 55_500 gas, but the user is passing 400_000 gas limit, they will be charged the `max(55_500, 320_000)`, so it would be 320_000. That is the value calculated after we return the 20% of the gasLimit back to the user, which is 80_000.
The logic compares the bigger of the two values - the gas consumed by the EVM and the 80% of the `gasLimit` that the user has passed.

In this way we lose track of the actual gas that was consumed by the EVM.

This HIP will add a new field `gas_consumed` to the `contract_result` table in the mirror node, which will contain the actually consumed gas by the EVM.

```

> create table if not exists contract_result

Reviewer: The mirror node database is not part of its public API and could change at anytime without another HIP, making this document become inaccurate. We should omit any reference to a specific table or column from this HIP and instead speak in general terms.

Reviewer 2: That's fair.
I believe this was my suggestion for insights into product changes.
Given it's not an external API it's more of a suggestion as part of the HIP so it's fine if the implementation is not exact.
If it's easier and better for accuracy though let's remove it to avoid confusion

If we sum up all of the `gasUsed` values of each ContractActions sidecar for a given transaction on top of the initial intrinsic gas, which is the initial gas, charged prior to the transaction execution, then we would get accurate value for the `gasConsumed` field.

## User stories

1. As a user, I want to see how much gas was consumed by the EVM during the execution of my transaction, when I retrieve a transaction info from the mirror-node REST APIs, so that I know if I can optimise my transaction cost
2. As a user, I want to see and compare how much gas I was charged for and how much gas was consumed by the EVM, when I inspect my transaction in hashscan

## Specification

In order to calculate the `gasUsed` field, the mirror node should have `CONTRACT_ACTION` and `CONTRACT_BYTECODE` sidecars enabled. Then, we would use an enhancement logic in the `mirror-node-importer`,
which will iterate over all `ContractAction` sidecar records for the transaction we want to import and calculate the `gasConsumed` field for.

> The new logic would have a mechanism to first calculate the intrinsic gas cost for the transaction being imported. We can inspect if the transaction has a `ContractByteCode` sidecar

Reviewer: This logic would not work for contract creates pre-sidecar. New mirror nodes could start and process all historical data so we have to handle this scenario. Same problem if sidecars are disabled. For those should we set gasConsumed null, zero, or equal to gasUsed?

Reviewer 2: Equal to gasUsed is probably best as null and 0 would just mean the caller would need to use gasUsed

Author: I updated the HIP with a description about setting the field as null. I think having `gasConsumed` equal to `gasUsed` in those cases, could be confusing, since we are telling the users that the EVM consumed as much gas as it was charged using the refund policy. This is often not the case and this case it would be a guess and we would mislead the user with returning wrong value for this field. If we have `null`, this would mean the value is absent and it's up to the users to decide how to use the gas information, if they need to handle it.

related to it. If yes, we have a contract creation transaction and we would calculate the intrinsic gas cost based on the length of the init bytecode. If we don't have such a sidecar,

> then the transaction represents a contract call and we would use the default intrinsic gas cost for contract calls, which is 21 000.

Reviewer: This logic does not discuss how to handle EthereumTransaction which can be either create or call.

Author: Based on the `consensusTimestamp` we could match specific sidecar files to the related transaction. If we find a `ContractBytecode` sidecar, then we have a create, if not - a call. So I think, we can still apply the same logic. I will add this clarification to the HIP.

> then the transaction represents a contract call and we would use the default intrinsic gas cost for contract calls, which is 21 000.

Reviewer: I have a concern about this hardcoded 21_000 intrinsic gas cost. Has it ever been something different in the past? Is it available in the record stream anywhere like 0.0.121? How would we keep this in sync with services and have different values over time?

Author: I don't think this value was ever changed in Hedera and it is described in the Ethereum yellow paper. That's the basic minimum cost we charge for any EVM operation. That's the cost of a basic transfer in Ethereum. I suspect if this value ever change it would be very rare and it shouldn't be an issue. Or we can just not point out the concrete value and just mention "intrinsic gas cost" in the HIP?

Reviewer: I'm fine with 21,000 in the HIP if it's never changed and is standard in Ethereum.

Reviewer 2: It's standard in Ethereum and all networks utilizing the EVM. 
Any change would require a new EIP and a new EVM version and likely HIP to support it.

> On top of the intrinsic gas cost, we would iterate over all `ContractAction`

Reviewer: There's no mention of whether historical data will be migrated to add the new data or not. I think we should add such a migration and call that out here.

Author: I agree that we should migrate historical contract results, whenever possible. I will add this description as well.

Author: Added description

sidecars and sum up the `gasUsed` field for each of them. The result would be the `gasConsumed` field for the transaction.

```
Example:

1. The transaction is a ContractCall

gas_consumed = intrinsic_gas + sum(contract_action_gas_used)

2. The transaction is a ContractCreate

> gas_consumed = intrinsic_gas + cost(contract_init_bytecode) + sum(contract_action_gas_used)

Reviewer: Does the intrisic_gas=21_000 for contract create as well or is it zero and just made up of initcode length?

Author: Yes, the intrinsic_gas is 21_000 for contract create as well, then we add a gas based on the size of the `initbytecode` that is being deployed. Additionally, we add the `contract action` gasUsed values in case we perform nested calls/creates inside the constructor.

```

The `gas_used` field in `contract_result` table would still have the same semantics and represent the gas that was charged to the user.

The following REST APIs will be extended with the new field:

```
1. GET /api/v1/contracts/results
2. GET /api/v1/contracts/results/{transactionIdOrHash}
3. GET /api/v1/contracts/{contractIdOrAddress}/results 
4. GET /api/v1/contracts/{contractIdOrAddress}/results/{timestamp}
```

The current response from these APIs looks like this:

```json
{

>   ...

Reviewer: The ellipses don't render properly. Maybe just add a note at the bottom that some fields omitted for brevity.

Author: Removed

  "error_message": "Out of gas",
  "failed_initcode": "0x856739",
  "from": "0000000000000000000000000000000000001f41",
  "function_parameters": "0xbb9f02dc6f0e3289f57a1f33b71c73aa8548ab8b",
  "gas_limit": 100000,
  "gas_price": "0x4a817c800",
  "gas_used": 320000,
  "hash": "0x3531396130303866616264653464",
  ...
}
```

> After the change, it will look like this:

Reviewer: Don't think you need to show the before and after. We don't normally do that for protos or APIs in HIPs. Can just say explicitly that a `gas_consumed` will be added to these APIs.

---
hip: 859
title: Add support for gasConsumed field in REST APIs related to contract results
author: Ivan Kavaldzhiev <ivan.kavaldzhiev@limechain.tech>, Stoyan Panayotov <stoyan.panayotov@limechain.tech>
working-group: Steven Sheehy <steven.sheehy@swirldslabs.com>, Nana Essilfie-Conduah <nana@swirldslabs.com>, Danno Ferrin <danno.ferrin@swirldslabs.com>, David Bakin <david.bakin@swirldslabs.com>
type: Standards Track
category: Mirror

> needs-council-approval: No

Reviewer 2: Change to yes, I was mistaken
```suggestion
needs-council-approval: No
```
```suggestion
needs-council-approval: Yes
```

Author: Fixed

status: Draft
created: 17.01.2024
discussions-to: https://github.com/hashgraph/hedera-improvement-proposal/discussions/860
---

## Abstract

> Currently, consensus nodes export a `gasUsed` field for each transaction, part of the record streaming. As a result, the `contract_result` table in the mirror node database contains a `gas_used` field for each transaction. This field value is calculated using the gas refund policy applied in the consensus node.

Author: Resolved, but I think `export` is a correct, since we have multiple consensus nodes in this description

> Currently, consensus nodes export a `gasUsed` field for each transaction, part of the record streaming. As a result, the `contract_result` table in the mirror node database contains a `gas_used` field for each transaction. This field value is calculated using the gas refund policy applied in the consensus node.

Reviewer 2: Actually you're right :)

This means that the value in this field represents how much gas the user was actually charged, not the amount of gas that the EVM consumed during the execution. This value can greatly differ from the actual amount that was used by the EVM. 

Let's say we use a 20% refund policy for the `gasLimit` that is sent by the user.

> That means if the EVM consumes for example 32 500 gas, but the user is passing 400 000 gas limit, they will be charged the `max(32 500, 320 000)`, so it would be 320 000. That is the value calculated after we return the 20% of the gasLimit back to the user, which is 80 000.

Author: Fixed

The logic compares the bigger of the two values - the gas consumed by the EVM and the 80% of the `gasLimit` that the user has passed.

In this way we lose track of the actual gas that was consumed by the EVM.

> This HIP will add a new field `gas_consumed` to the `contract_result` table in the mirror node, which will contain the actually consumed gas by the EVM.

Author: Fixed

## Motivation

Adding a `gas_consumed` field in the `contract_result` table would enrich user experience by providing more information about the transaction. 
This would allow users to better understand the gas refund policy and the actual gas consumption of their transactions. This would also allow users to understand how heavy their executed transaction was for the EVM.

## Rationale

> Consensus nodes export ContractActions sidecar records for each transaction after the introducing of [HIP-513](./hip-513.md). Each sidecar record contains a `gasUsed` field for the given action that the record represents (different type of call or contract creation).

Reviewer 2: ```suggestion
Consensus nodes export ContractActions sidecar records for each transaction after the introduction of [HIP-513](./hip-513.md). Each sidecar record contains a `gasUsed` field for the given action that the record represents (different type of call or contract creation).
```

Author: Fixed

## Specification

In order to calculate the `gasUsed` field, the mirror node should have `CONTRACT_ACTION` and `CONTRACT_BYTECODE` sidecars enabled. Then, we would use an enhancement logic in the `mirror-node-importer`,
which will iterate over all `ContractAction` sidecar records for the transaction we want to import and calculate the `gasConsumed` field for.

The new logic would have a mechanism to first calculate the intrinsic gas cost for the transaction being imported. We can inspect if the transaction has a `ContractByteCode` sidecar
related to it. If yes, we have a contract creation transaction and we would calculate the intrinsic gas cost based on the length of the init bytecode. If we don't have such a sidecar,
then the transaction represents a contract call and we would use the default intrinsic gas cost for contract calls, which is 21 000.

On top of the intrinsic gas cost, we would iterate over all `ContractAction`
sidecars and sum up the `gasUsed` field for each of them. The result would be the `gasConsumed` field for the transaction.

> The `gas_used` field in `contract_result` table would still have the same semantics and represent the gas that was charged to the user.

Reviewer 2: Let's express this in summary with an equation
e.g.
`gas_consumed = intrinsic_gas + sum(contract_action_gas_used) + cost(contract_init_bytecode)`

Author: Added examples

> 

Reviewer 2: Please add a list of all affected API endpoints and how the JSON response will change

Author: Added

## Backwards Compatibility

This HIP we will extend the REST APIs responses related to transaction info with an additional field `gas_consumed`. This won't be a breaking change, but the users should have in mind the introduction of the new field.

## Security Implications

None

## How to Teach This

Respective documentation will be added.

## Reference Implementation

Please follow [this issue](https://github.com/hashgraph/hedera-mirror-node/issues/7543)

## Rejected Ideas

While discussing the implementation of this HIP, we considered the following alternative:

> Adding a new `gasConsumed` field to the Record Stream V6 format, which would be a breaking change in the protobuf format and would require a new version of the Record Stream to be introduced.

Reviewer 2: I think there was another 1 or 2 ideas right?
One was calculating this all on the REST API side

Author: Added them as well

## References

> - [HIP-513](https://hips.hedera.com/hip/hip-513) - Smart Contract Traceability Extension

Reviewer 2: Also https://hips.hedera.com/hip/hip-226

Author: Added
